### DISCUSSION TITLE
추상 클래스와 인터페이스는 무슨 차이인가?

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#11] [Java]
.

**Comment by kunsanglee** [comment]
추상 클래스는 필드, 생성자, 메서드를 가질 수 있습니다. 추상 클래스에 선언된 모든 추상 메서드를 상속받는 하위 클래스에서 모두 구현해야 합니다.

인터페이스는 모든 메서드가 추상메서드이며, 필드와 생성자를 가질 수 없지만 상수는 가질 수 있습니다. 자바 8 부터는 default 메서드나, static 메서드를 포함할 수 있습니다. 인터페이스는 구현체가 인터페이스에 선언된 모든 메서드를 구현하도록 강제합니다.

중요한 차이점으로는 추상 클래스는 단일 상속만 가능하지만, 인터페이스는 다중 상속이 가능합니다. 추상 클래스는 상속받는 클래스에 공통된 구현을 제공하는데 유리하고, 인터페이스는 다양한 클래스가 동일한 인터페이스를 구현하도록 하는 데 유리합니다.

**Reply by alstn113** [reply]
첫 번째 줄에서 최소 하나의 추상 메서드를 가져야한다고 했는데 추상 메서드가 없어도 추상 클래스를 만들 수 있습니다.

**Comment by seunghye218** [comment]
추상 클래스는 인스턴스화할 수 없고 추상 메소드와 일반 메소드를 모두 포함할 수 있는 클래스입니다. 공통적인 기능을 미리 구현해 두어 상속받는 클래스에서 이를 재사용할 수 있게 합니다. 이는 코드의 중복을 줄이고 유지보수를 용이하게 하는 데 도움이 됩니다.

인터페이스는 클래스가 구현해야 하는 메서드 집합을 지정하는 계약입니다.
인터페이스는 다중 상속을 필요로 하는 상황에서 여러 인터페이스를 구현함으로써 한 클래스가 다양한 역할을 수행할 수 있게 합니다.

코드의 재사용이 중요한 경우나 공통적인 기능을 미리 정의해 두고 싶은 경우에는 추상 클래스를 사용하는 것이 적합합니다.
반면, 다중 상속이 필요하거나 구현 객체 간의 계약을 정의할 필요가 있는 경우 인터페이스를 사용하는 것이 좋습니다.

**Comment by le2sky** [comment]
https://f-lab.kr/insight/understanding-java-interfaces-and-abstract-classes?gad_source=1&gclid=CjwKCAjw_e2wBhAEEiwAyFFFoz9hYe0eZzKMA-0zElI3NuZISe3yraQH--5Po_iMug9skcZH-FeceRoCCmQQAvD_BwE

**Comment by GIVEN53** [comment]
추상 클래스
일반 + 추상 메서드
필드, 생성자
상속 재사용, 코드 중복 제거

인터페이스
추상 메서드
상수
Java 8 default, static
다중 상속, 동일한 동작 구현
